File: Tugas Akhir Semester 1/TrashBinBe/app/Models/MarketplaceListing.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class MarketplaceListing extends Model
{
    use HasFactory;

    /**
     * The attributes that are mass assignable.
     *
     * @var array<int, string>
     */
    protected $fillable = [
        'seller_id',
        'waste_category_id',
        'title',
        'description',
        'quantity',
        'unit',
        'price_per_unit',
        'total_price',
        'condition',
        'location',
        'lat',
        'lng',
        'status',
        'photos',
        'views_count',
        'expires_at',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'lat' => 'decimal:8',
        'lng' => 'decimal:8',
        'quantity' => 'decimal:2',
        'price_per_unit' => 'decimal:2',
        'total_price' => 'decimal:2',
        'photos' => 'array',
        'expires_at' => 'datetime',
        'views_count' => 'integer',
    ];

    /**
     * Get the seller of this listing
     */
    public function seller()
    {
        return $this->belongsTo(User::class, 'seller_id');
    }

    /**
     * Get the waste category for this listing
     */
    public function wasteCategory()
    {
        return $this->belongsTo(WasteCategory::class, 'waste_category_id');
    }

    /**
     * Get the orders for this listing
     */
    public function orders()
    {
        return $this->hasMany(Order::class);
    }
    
    /**
     * Scope to get available listings
     */
    public function scopeAvailable($query)
    {
        return $query->where('status', 'available');
    }
    
    /**
     * Scope to get expired listings
     */
    public function scopeExpired($query)
    {
        return $query->where('status', 'expired')->orWhere('expires_at', '<', now());
    }
    
    /**
     * Scope to get sold listings
     */
    public function scopeSold($query)
    {
        return $query->where('status', 'sold');
    }
    
    /**
     * Scope to get reserved listings
     */
    public function scopeReserved($query)
    {
        return $query->where('status', 'reserved');
    }
}

File: Tugas Akhir Semester 1/TrashBinFe/app/src/main/java/com/trashbin/app/data/model/MarketplaceListing.kt
package com.trashbin.app.data.model

import com.google.gson.annotations.SerializedName

data class MarketplaceListing(
    @SerializedName("id") val id: Int,
    @SerializedName("seller_id") val sellerId: Int,
    @SerializedName("waste_category_id") val categoryId: Int,
    @SerializedName("title") val title: String,
    @SerializedName("description") val description: String,
    @SerializedName("quantity") val quantity: Double,
    @SerializedName("unit") val unit: String,
    @SerializedName("price_per_unit") val pricePerUnit: Double,
    @SerializedName("total_price") val totalPrice: Double,
    @SerializedName("condition") val condition: String, // clean, needs_cleaning, mixed
    @SerializedName("location") val location: String,
    @SerializedName("lat") val lat: String,
    @SerializedName("lng") val lng: String,
    @SerializedName("status") val status: String,
    @SerializedName("photos") val photos: List<String>,
    @SerializedName("views_count") val views: Int,
    @SerializedName("expires_at") val expiresAt: String?,
    @SerializedName("waste_category") val category: WasteCategory,
    @SerializedName("seller") val seller: SellerInfo,
    @SerializedName("created_at") val createdAt: String?,
    @SerializedName("updated_at") val updatedAt: String?
)

data class SellerInfo(
    @SerializedName("id") val id: Int,
    @SerializedName("name") val name: String,
    @SerializedName("avatar") val avatar: String?,
    @SerializedName("points") val points: Int
)